### PR TITLE
Hw05 tests

### PR DESCRIPTION
**Pull request by TiBulat**

### CODE REVIEW

File: HW03/tim_bulat/spec/tim_bulat_hw_03_t_03_rspec.rb [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+      2018-04-23 17:18:59.8 ubuntu-xenial[14319] Debug - inside docker_handle_event'
+    end
+
+    let(:log_no_core) do
+      '10.6.246.103 - - [23/Apr/2018:20:30:39 +0300] "POST /test/2/messages HTTP/1.1" 200 48 0.0498
+              10.6.246.101 - - [23/Apr/2018:20:30:42 +0300] "POST /test/2/run HTTP/1.1" 200 - 0.2277
+              2018-04-23 20:30:42: SSL ERROR, peer: 10.6.246.101, peer cert: , #<Puma::MiniSSL::SSL: System error: Undefined error: 0 - 0>
+              10.6.246.101 - - [23/Apr/2018:20:31:39 +0300] "POST /test/2/messages HTTP/1.1" 200 48 0.0290'
+    end
+    let(:log_more_than_two) do
```
Comment (DashaSledopit) [inline HW03/tim_bulat/spec/tim_bulat_hw_03_t_03_rspec.rb:23]:
```suggestion
    end

    let(:log_more_than_two) do
```

File: HW03/tim_bulat/spec/tim_bulat_hw_03_t_03_rspec.rb [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    context 'when two core calling' do
+      subject { described_class.new.task3(log) }
+
+      let(:output) { '49.1' }
+
+      it 'returns string' do
+        expect(subject).to be_a String
+      end
+
+      it 'returns strng with calculate time ' do
```
Comment (DashaSledopit) [inline HW03/tim_bulat/spec/tim_bulat_hw_03_t_03_rspec.rb:70]:
```suggestion
      it 'returns string with calculated time ' do
```

File: HW03/tim_bulat/spec/tim_bulat_hw_03_t_04_rspec.rb [hunk 3]
```diff
@@ -0,0 +1,55 @@
+# frozen_string_literal: true
+
+require_relative '../tim_bulat_hw_03_t_04'
+
+RSpec.describe Homework3 do
+  describe '#task4' do
+    context 'when empty string' do
+      subject { described_class.new.task4('') }
+
+      it 'expcet to return hash' do
```
Comment (DashaSledopit) [inline HW03/tim_bulat/spec/tim_bulat_hw_03_t_04_rspec.rb:10]:
```suggestion
      it 'returns Hash' do
```

File: HW03/tim_bulat/spec/tim_bulat_hw_03_t_04_rspec.rb [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        expect(subject).to be_a(Hash)
+      end
+
+      it 'expcet to return hash with 0s' do
```
Comment (DashaSledopit) [inline HW03/tim_bulat/spec/tim_bulat_hw_03_t_04_rspec.rb:14]:
```suggestion
      it 'returns hash with 0s' do
```

File: HW03/tim_bulat/spec/tim_bulat_hw_03_t_04_rspec.rb [hunk 5]
```diff
@@ -0,0 +1,55 @@
+# frozen_string_literal: true
+
+require_relative '../tim_bulat_hw_03_t_04'
+
+RSpec.describe Homework3 do
+  describe '#task4' do
```
Comment (DashaSledopit) [inline HW03/tim_bulat/spec/tim_bulat_hw_03_t_04_rspec.rb:6]: you can create 1 subject here :  `subject { described_class.new.task4(string) }` and change string with the help of `let` in each context

File: HW04/tim_bulat/spec/mentor_spec.rb [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+require_relative '../lib/student'
+require_relative '../lib/mentor'
+require_relative '../lib/homework'
+require_relative '../lib/notification'
+
+RSpec.describe Mentor do
+  let(:student) { Student.new(name: 'students_name', surname: 'students_surname') }
+  let(:mentor) { described_class.new(name: 'mentor_name', surname: 'mentor_surname') }
+  let(:hw) { Homework.new(title: 'title', description: 'description', student: student, mentor: mentor) }
```
Comment (DashaSledopit) [inline HW04/tim_bulat/spec/mentor_spec.rb:11]:
```suggestion
  let(:homework) { Homework.new(title: 'title', description: 'description', student: student, mentor: mentor) }
```

File: HW04/tim_bulat/spec/mentor_spec.rb [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+      expect(mentor.name).to eq 'mentor_name'
+    end
+
+    it 'returns correct surname' do
+      expect(mentor.surname).to eq 'mentor_surname'
+    end
+  end
+
+  context 'when subscribed to student' do
+    it 'student added' do
```
Comment (DashaSledopit) [inline HW04/tim_bulat/spec/mentor_spec.rb:24]: you should describe in `it block` what the method does. the line 31 is a good example

File: HW04/tim_bulat/spec/mentor_spec.rb [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+      mentor.subscribe_to!(student)
+      expect(mentor.students.empty?).to eq false
+    end
+  end
+
+  describe '#add_homework!' do
+    it 'returs Homework object' do
+      homework = mentor.add_homework(title: 'title', description: 'description', student: student)
```
Comment (DashaSledopit) [inline HW04/tim_bulat/spec/mentor_spec.rb:32]: don't create variables this way

File: HW04/tim_bulat/spec/student_spec.rb [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+  describe '#to_check!' do
+    it 'changes status of homework' do
+      student.to_check!(hw)
+      expect(hw.status).to eq 'to check'
+    end
+
+    it 'notifies mentor' do
+      student.to_check!(hw)
+      expect(mentor.notes.empty?).to eq false
```
Comment (DashaSledopit) [inline HW04/tim_bulat/spec/student_spec.rb:38]:
```suggestion
      expect(mentor.notes).to be_empty
```

File: HW04/tim_bulat/spec/student_spec.rb [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+
+  context 'when created' do
+    it 'returns correct name' do
+      expect(student.name).to eq 'students_name'
+    end
+
+    it 'returns correct surname' do
+      expect(student.surname).to eq 'students_surname'
+    end
+  end
```
Comment (DashaSledopit) [inline HW04/tim_bulat/spec/student_spec.rb:21]: unnecessary

File: HW03/tim_bulat/spec/tim_bulat_hw_03_t_03_rspec.rb [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+    context 'when two core calling' do
+      subject { described_class.new.task3(log) }
+
+      let(:output) { '49.1' }
+
+      it 'returns string' do
+        expect(subject).to be_a String
+      end
+
+      it 'returns stirng with calculated time ' do
```
Comment (DashaSledopit) [inline HW03/tim_bulat/spec/tim_bulat_hw_03_t_03_rspec.rb:71]: be more careful please
```suggestion
      it 'returns string with calculated time ' do
```

File: HW04/tim_bulat/spec/mentor_spec.rb [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+require_relative '../lib/mentor'
+require_relative '../lib/homework'
+require_relative '../lib/notification'
+
+RSpec.describe Mentor do
+  let(:student) { Student.new(name: 'students_name', surname: 'students_surname') }
+  let(:mentor) { described_class.new(name: 'mentor_name', surname: 'mentor_surname') }
+  let(:homework) { Homework.new(title: 'title', description: 'description', student: student, mentor: mentor) }
+
+  context 'when subscribed to student' do
```
Comment (DashaSledopit) [inline HW04/tim_bulat/spec/mentor_spec.rb:13]: have you missed describe block?

File: HW04/tim_bulat/spec/student_spec.rb [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+
+RSpec.describe Student do
+  let(:student) { described_class.new(name: 'students_name', surname: 'students_surname') }
+  let(:mentor) { Mentor.new(name: 'mentor_name', surname: 'mentor_surname') }
+  let(:homework) { Homework.new(title: 'title', description: 'description', student: student, mentor: mentor) }
+
+  describe '#to_work!' do
+    it 'notifies mentor' do
+      student.to_work!(homework)
+      expect(mentor.notes.empty?).to eq false
```
Comment (DashaSledopit) [inline HW04/tim_bulat/spec/student_spec.rb:16]: could you remake all similar checks as in the previous class?

File: HW04/tim_bulat/spec/mentor_spec.rb [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+  describe '#subscribe_to!' do
+    context 'when subscribed to student' do
+      it 'add student to mentors list' do
+        mentor.subscribe_to!(student)
+        expect(mentor.students).not_to be_empty
+      end
+    end
+  end
+
+  describe '#add_homework!' do
```
Comment (DashaSledopit) [inline HW04/tim_bulat/spec/mentor_spec.rb:22]: does this method also adds any notifications?

File: HW04/tim_bulat/spec/mentor_spec.rb [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+    it 'returs Homework object' do
```
Comment (DashaSledopit) [inline HW04/tim_bulat/spec/mentor_spec.rb:23]:
```suggestion
    it 'returns Homework object' do
```

**Comment by TiBulat** [comment]
Really sorry for typos. I'm becoming dyslexic when it comes to english

File: HW04/tim_bulat/spec/mentor_spec.rb [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+  describe '#subscribe_to!' do
+    context 'when subscribed to student' do
+      it 'adds needed student to list' do
+        mentor.subscribe_to!(student)
+        expect(mentor.students).to include student
+      end
+    end
+  end
+
+  describe '#add_homework!' do
```
Comment (DashaSledopit) [inline HW04/tim_bulat/spec/mentor_spec.rb:22]: you can create a subject here

File: HW03/tim_bulat/spec/tim_bulat_hw_03_t_01_rspec.rb [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+
+    context 'when log has NO errors' do
+      subject { described_class.new.task1(log_no_errors) }
+
+      it 'returns empty string' do
+        expect(subject).to eq ''
+      end
+    end
+
+    context 'when log has erorr lines' do
```
Comment (DashaSledopit) [inline HW03/tim_bulat/spec/tim_bulat_hw_03_t_01_rspec.rb:39]:
```suggestion
    context 'when log has error lines' do
```

File: HW03/tim_bulat/spec/tim_bulat_hw_03_t_02_rspec.rb [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+
+    context 'when wrong format' do
+      subject { described_class.new.task2(log_no_post) }
+
+      it 'return empty array' do
+        expect(subject).to eq []
+      end
+    end
+
+    context 'when string is right' do
```
Comment (DashaSledopit) [inline HW03/tim_bulat/spec/tim_bulat_hw_03_t_02_rspec.rb:43]: could you please make the context more specific?

File: HW04/tim_bulat/lib/mentor.rb [hunk 19]
```diff
@@ -8,6 +8,8 @@ def initialize(name:, surname:)
     @students = []
   end
 
+  attr_reader :students
```
Comment (DashaSledopit) [inline HW04/tim_bulat/lib/mentor.rb:11]: move it up above the initializer, please

File: HW04/tim_bulat/spec/mentor_spec.rb [hunk 20]
```diff
@@ -0,0 +1,56 @@
...
+
+  describe '#reject_to_work!' do
+    it 'changes status of homework to rejected' do
+      mentor.reject_to_work!(homework)
+      expect(homework.status).to eq 'rejected'
+    end
+
+    it 'notifies student' do
+      mentor.reject_to_work!(homework)
+      expect(student.notes).not_to be_empty
```
Comment (DashaSledopit) [inline HW04/tim_bulat/spec/mentor_spec.rb:41]: let's use expect {}.to change {}.from().to() everywhere

File: HW03/tim_bulat/spec/tim_bulat_hw_03_t_01_rspec.rb [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+  describe '#task1' do
+    let(:log_with_errors) do
+      '10.6.246.103 - - [23/Apr/2018:20:30:39 +0300] "POST /grid/2/messages HTTP/1.1" 200 48 0.0498
+      10.6.246.101 - - [23/Apr/2018:20:30:42 +0300] "POST /grid/2/event HTTP/1.1" 200 - 0.2277
+      2018-04-23 20:30:42: SSL error, peer: 10.6.246.101, peer cert: , #<Puma::MiniSSL::SSLError: System error: Undefined error: 0 - 0>
+      2018-04-23 21:30:42: SSL error
+      10.6.246.101 - - [23/Apr/2018:20:29:39 +0300] "POST /grid/2/messages HTTP/1.1" 200 48 0.0290'
+    end
+
+    let(:log_no_errors) do
```
Comment (DashaSledopit) [inline HW03/tim_bulat/spec/tim_bulat_hw_03_t_01_rspec.rb:15]:
```suggestion
    let(:log_without_errors) do
```

File: HW03/tim_bulat/spec/tim_bulat_hw_03_t_01_rspec.rb [hunk 22]
```diff
@@ -0,0 +1,47 @@
...
+      '10.6.246.103 - - [23/Apr/2018:20:30:39 +0300] "POST /grid/2/messages HTTP/1.1" 200 48 0.0498
+        10.6.246.101 - - [23/Apr/2018:20:30:42 +0300] "POST /grid/2/event HTTP/1.1" 200 - 0.2277
+        10.6.246.101 - - [23/Apr/2018:20:29:39 +0300] "POST /grid/2/messages HTTP/1.1" 200 48 0.0290'
+    end
+
+    let(:error) { '2018-04-23 20:30:42: SSL error, peer: 10.6.246.101, peer cert: , #<Puma::MiniSSL::SSLError: System error: Undefined error: 0 - 0>' }
+
+    context 'when empty string' do
+      subject { described_class.new.task1('') }
```
Comment (DashaSledopit) [inline HW03/tim_bulat/spec/tim_bulat_hw_03_t_01_rspec.rb:24]: let's move the subject from all contexts and define onlu logs here

File: HW03/tim_bulat/spec/tim_bulat_hw_03_t_02_rspec.rb [hunk 23]
```diff
@@ -0,0 +1,55 @@
...
+    let(:formated_strings) do
+      [
+        'DATE: 23/Apr/2018:20:30:39 +0300 FROM: 10.6.246.103 TO: /TEST/2/MESSAGES',
+        'DATE: 23/Apr/2018:20:30:42 +0300 FROM: 10.6.246.101 TO: /TEST/2/RUN',
+        'DATE: 23/Apr/2018:20:31:39 +0300 FROM: 10.6.246.101 TO: /TEST/2/MESSAGES'
+      ]
+    end
+
+    context 'when empty string' do
+      subject { described_class.new.task2('') }
```
Comment (DashaSledopit) [inline HW03/tim_bulat/spec/tim_bulat_hw_03_t_02_rspec.rb:28]: the same thing about the subject

File: HW03/tim_bulat/spec/tim_bulat_hw_03_t_02_rspec.rb [hunk 24]
```diff
@@ -0,0 +1,55 @@
+# frozen_string_literal: true
+
+require_relative '../tim_bulat_hw_03_t_02'
+
+RSpec.describe Homework3 do
+  describe '#task2' do
+    let(:log_no_post) do
```
Comment (DashaSledopit) [inline HW03/tim_bulat/spec/tim_bulat_hw_03_t_02_rspec.rb:7]:
```suggestion
    let(:log_without_post) do
```

File: HW03/tim_bulat/spec/tim_bulat_hw_03_t_02_rspec.rb [hunk 25]
```diff
@@ -0,0 +1,55 @@
...
+
+    context 'when empty string' do
+      subject { described_class.new.task2('') }
+
+      it 'returns empty array' do
+        expect(subject).to eq []
+      end
+    end
+
+    context 'when wrong format' do
```
Comment (DashaSledopit) [inline HW03/tim_bulat/spec/tim_bulat_hw_03_t_02_rspec.rb:35]: what are checking here? when there is no post methods in the log

File: HW03/tim_bulat/spec/tim_bulat_hw_03_t_03_rspec.rb [hunk 26]
```diff
@@ -0,0 +1,98 @@
+# frozen_string_literal: true
+
+require_relative '../tim_bulat_hw_03_t_03'
+
+RSpec.describe Homework3 do
```
Comment (DashaSledopit) [inline HW03/tim_bulat/spec/tim_bulat_hw_03_t_03_rspec.rb:5]: 1. the same thing with the subject
2. let's try to come up with more specific names for the variables

File: HW03/tim_bulat/spec/tim_bulat_hw_03_t_03_rspec.rb [hunk 27]
```diff
@@ -0,0 +1,98 @@
...
+        2018-04-23 17:17:49.7 ubuntu-xenial[14319] Debug - connecting to: 10.6.246.101
+        2018-04-23 17:17:49.8 ubuntu-xenial[14319] Debug - docker event processed
+        2018-04-23 17:18:19.5 ubuntu-xenial[14319] Debug - monitoring grid communication health
+        2018-04-23 17:18:38.8 ubuntu-xenial[14319] Debug - Calling core with action: messages
+        2018-04-23 17:18:38.8 ubuntu-xenial[14319] Debug - connecting to: 10.6.246.101
+        2018-04-23 17:18:59.8 ubuntu-xenial[14319] Debug - inside docker_handle_event
+        2018-04-23 15:19:38.8 ubuntu-xenial[14319] Debug - Calling core with action: messages'
+    end
+
+    context 'when empty string' do
```
Comment (DashaSledopit) [inline HW03/tim_bulat/spec/tim_bulat_hw_03_t_03_rspec.rb:46]:
```suggestion
    context 'when log is empty' do
```

File: HW03/tim_bulat/spec/tim_bulat_hw_03_t_03_rspec.rb [hunk 28]
```diff
@@ -0,0 +1,98 @@
...
+      subject { described_class.new.task3('') }
+
+      it 'returns 0' do
+        expect(subject).to eq '0'
+      end
+    end
+
+    context 'when no core actions' do
```
Comment (DashaSledopit) [inline HW03/tim_bulat/spec/tim_bulat_hw_03_t_03_rspec.rb:54]:
```suggestion
    context 'when there are no core actions' do
```

File: HW03/tim_bulat/spec/tim_bulat_hw_03_t_02_rspec.rb [hunk 29]
```diff
@@ -0,0 +1,57 @@
...
+
+      it 'returns empty array' do
+        expect(subject).to eq []
+      end
+    end
+
+    context 'when there are not post methods in the log' do
+      let(:log) { log_without_post }
+
+      it 'return empty array' do
```
Comment (DashaSledopit) [inline HW03/tim_bulat/spec/tim_bulat_hw_03_t_02_rspec.rb:40]:
```suggestion
      it 'returns empty array' do
```

File: HW04/tim_bulat/spec/mentor_spec.rb [hunk 30]
```diff
@@ -0,0 +1,55 @@
...
+
+    it 'notifies student' do
+      expect { subject }.to change { student.notes.size }.from(0).to(1)
+    end
+  end
+
+  describe '#reject_to_work!' do
+    it 'changes status of homework to rejected' do
+      mentor.reject_to_work!(homework)
+      expect(homework.status).to eq 'rejected'
```
Comment (DashaSledopit) [inline HW04/tim_bulat/spec/mentor_spec.rb:37]: you can also use` to_change` here

File: HW04/tim_bulat/spec/mentor_spec.rb [hunk 31]
```diff
@@ -0,0 +1,55 @@
...
+
+    it 'notifies student' do
+      expect { mentor.reject_to_work!(homework) }.to change { student.notes.size }.from(0).to(1)
+    end
+  end
+
+  describe '#accept!' do
+    it 'changes status of homework to accepted' do
+      mentor.accept!(homework)
+      expect(homework.status).to eq 'accepted'
```
Comment (DashaSledopit) [inline HW04/tim_bulat/spec/mentor_spec.rb:48]: and here

Comment (DashaSledopit) [reply HW04/tim_bulat/spec/mentor_spec.rb:48]: and in all changes of homework's status

**Pull request merged**